builtin/credential/aws/path_role.go

```diff
@@ -241,6 +282,16 @@ func (b *backend) nonLockedAWSRole(s logical.Storage, roleName string) (*awsRole
 		}
 	}
 
+	// Check if there was no pre-existing AuthType set (from older versions)
```

Comment:
A lock upgrade primitive would be great, yes, but I can imagine how it can get very hard to write one. Also, I was very much aware of the race window between the read and the writes. We release the read lock explicitly instead of the defer statement, upgrade the role entry, and invoke `lockedSetAWSRole` which acquires the write lock before updating the role. This sure is better than modifying the role with a read lock. A stale read is better than an inconsistent write :-)